### DISCUSSION TITLE
Implementing CIS Benchmarks for Iptables

### DISCUSSION THREAD

**Discussion by ozmodiar192** [bottlerocket-os/bottlerocket#2734] [General]
Hello,
we are starting to implement the CIS benchmarks for Bottlerocket.  The CIS Benchmark document shows an example shell script with iptables commands like this:
```
# Ensure default deny firewall policy
iptables -P INPUT DROP
iptables -P OUTPUT DROP
iptables -P FORWARD DROP
```
Then it says:
```
The above script can be deployed through a bootstrap container, which is the
mechanism that Bottlerocket provides for running custom code during the boot process.
```
It is my understanding that executing IPTables commands in a bootstrap container to modify rules would not work.  Is the CIS Benchmark misleading, or am I misunderstanding the nature of bootstrap containers?  My understanding is that the Bottlerocket way of handing this would be to use the bootstrap container to write out rules to the appropriate file system location like /etc/sysconfig/iptables using the root file system mount.

**Comment by arnaldo2792** [comment]
Hi @ozmodiar192, after [this change] was merged bootstrap containers gained the capabilities required to modify iptables rules without requiring modifications in the root filesystem. Keep in mind that you will want to configure your bootstrap container as `mode = always` in order to apply the iptables changes on subsequent reboots.

[this change]: https://github.com/bottlerocket-os/bottlerocket/pull/2266

**Reply by ozmodiar192** [reply]
Thank you, that's super helpful.  I really appreciate it, and thanks for all the work you do!

**Reply by nike21oct** [reply]
Hi , I also want to Implementing CIS Benchmarks for Iptables , but i want to implement with user_data of terraform so what could be the options to implement IPtables rule with the help of settings?
user_data = base64encode(<<EOT
[settings.bootstrap-containers.admin]
mode = "once"
user-data = """
#!/bin/bash
iptables -P INPUT DROP
iptables -P OUTPUT DROP
iptables -P FORWARD DROP
iptables -A INPUT -i lo -j ACCEPT
iptables -A OUTPUT -o lo -j ACCEPT
iptables -A INPUT -s 127.0.0.0/8 -j DROP
"""
EOT
  )
is this correct way?

**Reply by stmcginnis** [reply]
Hi @nike21oct . The admin container and bootstrap container are two different concepts. Both are only run if configured to do so, but the admin container is the only one that is available by default with just a configuration setting to enable it. It allows SSH access and access other troubleshooting tools. It does not have a mechanism to run arbitrary scripts on start up.

More details about the admin container [can be found here](https://github.com/bottlerocket-os/bottlerocket-admin-container).

The bootstrap container is the only one guaranteed to run early at initialization that will allow you to make changes to the system. But you need to provide a container image to run as this bootstrap container. For the case of configuring iptables rules for Bottlerocket CIS Benchmark level 2 compliance, you would need to set up a bootstrap container to make these changes at boot time. A good example of doing this can be found in this workshop: https://catalog.us-east-1.prod.workshops.aws/workshops/165b0729-2791-4452-8920-53b734419050/en-US/10-regulatory-compliance/cis-bottlerocket-eks/build-bootstrap-image

On the terraform side, you would just need to pass [the user data to enable the bootstrap container](https://github.com/bottlerocket-os/bottlerocket#bootstrap-containers-settings).

**Reply by nike21oct** [reply]
Hi @stmcginnis , thanks for your reply, I followed below configuration via terraform but it works for sysctl but does not work for IPtables rule , so is there any mistake in my configuration synatx , can you please help me here

user_data = base64encode(<<EOT
[settings.kernel.sysctl] 
"net.ipv4.conf.all.send_redirects" = "0" 
"net.ipv4.conf.default.send_redirects" = "0" 
"net.ipv4.conf.all.accept_redirects" = "0" 
"net.ipv4.conf.default.accept_redirects" = "0" 
"net.ipv6.conf.all.accept_redirects" = "0" 
"net.ipv6.conf.default.accept_redirects" = "0" 
"net.ipv4.conf.all.secure_redirects" = "0" 
"net.ipv4.conf.default.secure_redirects" = "0" 
"net.ipv4.conf.all.log_martians" = "1" 
"net.ipv4.conf.default.log_martians" = "1" 

[settings.bootstrap-containers.cis-bootstrap]
source = "registry-path/bottlerocket-benchmarking/bootstrap"
mode = "once"
EOT
  )

**Reply by stmcginnis** [reply]
Correct, I don't believe iptables rules can be configured via the sysctl calls.

**Reply by nike21oct** [reply]
no no sysctl is used for changing the kernel settings but this one [settings.bootstrap-containers.cis-bootstrap] is having source which is having bootstrap container image as described in the document https://catalog.us-east-1.prod.workshops.aws/workshops/165b0729-2791-4452-8920-53b734419050/en-US/10-regulatory-compliance/cis-bottlerocket-eks/build-bootstrap-image. you shared , I build the image and stored in source , it runs via terraform but did not change the  iptables rule

**Reply by stmcginnis** [reply]
Can you share your `Dockerfile` and script for this bootstrap container?

Make sure whatever "registry-path/bottlerocket-benchmarking/bootstrap" actually points to is accessible.

You will likely want `mode = "always"` so the iptables settings are applied every time the node boots.

**Reply by nike21oct** [reply]
hi @stmcginnis , i used the docker file provided in this URL for creating bootstrap container "https://github.com/aws-samples/containers-blog-maelstrom/tree/main/cis-bottlerocket-benchmark-eks/bottlerocket-cis-bootstrap-image"  and store the image in my private registry to gitlab and used below setting in user_data.

user_data = base64encode(<<EOT`
[settings.bootstrap-containers.cis-bootstrap]
source = "registry.devops.telekom.de/mavie/system/tools/bottlerocket-benchmarking/bootstrap"
mode = "always"
EOT
)

can you please guide what could be the issue ?

**Reply by stmcginnis** [reply]
Nothing obvious is standing out to me. Maybe try connecting (`aws ssm start-session --target $INSTANCE_ID` or SSH if set up), `enter-admin-container` to connect, the `sudo sheltie` to get a prompt as if you are on the OS itself. From there, you can try `journalctl | grep boot` and see if there are any error messages from the initialization that would point to what the failure might be.

If it is successful, you should see something like:

```
Aug 07 16:47:50 ip-192-168-42-247.us-east-2.compute.internal bootstrap-containers[1277]: 16:47:50 [INFO] bootstrap-containers started
Aug 07 16:47:50 ip-192-168-42-247.us-east-2.compute.internal bootstrap-containers[1277]: 16:47:50 [INFO] Mode for 'bootstrap' is 'always'
Aug 07 16:47:50 ip-192-168-42-247.us-east-2.compute.internal systemd[1]: Finished bootstrap container bootstrap.
```

For my case, here are all the steps I did to double check functionality.

I created a container using this Dockerfile:

```dockerfile
FROM alpine
COPY ./bootstrap-script.sh /
RUN chmod +x /bootstrap-script.sh
RUN apk update && apk add bash && apk add iptables && apk add ip6tables
ENTRYPOINT ["/bootstrap-script.sh"]
```

That pulls in a simple script that I just made one small iptables change to see it working:

```bash
#!/usr/bin/env bash

# Flush iptables rules
iptables -F

iptables -I INPUT -p tcp -m tcp --dport 22222 -j ACCEPT
```

I built that container image and pushed it to my private registry.

**One thing to note**: I originally did have an issue here, so maybe double check if this might apply to you. My development machine is an ARM64 box, but my cluster nodes are AMD64. That caused an "exec format error" when it tried to run it.

I then launched instances with the following user data:

```toml
[settings.bootstrap-containers.bootstrap]
source = "id.dkr.ecr.us-east-2.amazonaws.com/bootstrap:latest"
mode = always
```

Made sure there were no errors in the system log, and double checked the iptables configuration was applied:

```sh
bash-5.1# iptables -L | grep 22             
ACCEPT     tcp  --  anywhere             anywhere             tcp dpt:22222
```

The only thing I see different is there is no tag specified in your user data. The logs should have some message if that is the cause of the error.

**Reply by arnaldo2792** [reply]
Yes, you need to provide the tag otherwise `host-ctr` (our client that runs the bootstrap containers) will fail to pull down the container image. FWIW, you can use this from the admin container to only get the bootstrap container's logs: `journalctl -D /.bottlerocket/rootfs/var/log -u bootstrap-containers@bootstrap.service`

You can use `bootstrap-containers@<container-name-in-API>.service` to filter out logs from the journal :+1:

**Comment by nike21oct** [comment]
hi @stmcginnis and all, i have a EKS cluster which is using bottlerocket AMI and nginx as a ingress controller and when i implemented these IP tables rules by bootstrap-container my application stop open from outside the cluster  , i mean my ingress is not functioning my nginx ingress controller pod went to crashloopbackoff , ngix controller load balancer in aws in target group the Protocol : Port is TLS: 32443 and health check is using protocol http and port is 32002, so what should i need to do?
please help me here

**Comment by arnaldo2792** [comment]
From the CIS benchmark:

> With a default accept policy the firewall will accept any packet that is not configured to
be denied. It is easier to allow list acceptable usage than to deny list unacceptable
usage.

Thus, you need to manually configure the rules you need for your application to work.

**Reply by nike21oct** [reply]
hi @arnaldo2792 , below is the iptables rules which i am using in my script and created image from this and use while botting up the nodes.

#!/usr/bin/env bash

# Flush iptables rules
iptables -F

# 3.4.1.1 Ensure IPv4 default deny firewall policy (Automated)
iptables -P INPUT DROP
iptables -P OUTPUT DROP
iptables -P FORWARD DROP

# Allow inbound traffic for kubelet (so kubectl logs/exec works)
iptables -I INPUT -p tcp -m tcp --dport 10250 -j ACCEPT

#Adding nodeport of nginx ingress controlleer 
iptables -I INPUT -p tcp -m tcp --dport 32443 -j ACCEPT  # For TLS traffic
iptables -I INPUT -p tcp -m tcp --dport 32002 -j ACCEPT  # For Health Checks
iptables -I INPUT -p tcp -m tcp --dport 32080 -j ACCEPT

# 3.4.1.2 Ensure IPv4 loopback traffic is configured (Automated)
iptables -A INPUT -i lo -j ACCEPT
iptables -A OUTPUT -o lo -j ACCEPT
iptables -A INPUT -s 127.0.0.0/8 -j DROP

# 3.4.1.3 Ensure IPv4 outbound and established connections are configured (Manual)
iptables -A OUTPUT -p tcp -m state --state NEW,ESTABLISHED -j ACCEPT
iptables -A OUTPUT -p udp -m state --state NEW,ESTABLISHED -j ACCEPT
iptables -A OUTPUT -p icmp -m state --state NEW,ESTABLISHED -j ACCEPT
iptables -A INPUT -p tcp -m state --state ESTABLISHED -j ACCEPT
iptables -A INPUT -p udp -m state --state ESTABLISHED -j ACCEPT
iptables -A INPUT -p icmp -m state --state ESTABLISHED -j ACCEPT


# Flush ip6tables rules 
ip6tables -F

# 3.4.2.1 Ensure IPv6 default deny firewall policy (Automated)
ip6tables -P INPUT DROP
ip6tables -P OUTPUT DROP
ip6tables -P FORWARD DROP

# Allow inbound traffic for kubelet on ipv6 if needed (so kubectl logs/exec works)
ip6tables -A INPUT -p tcp --destination-port 10250 -j ACCEPT

# 3.4.2.2 Ensure IPv6 loopback traffic is configured (Automated)
ip6tables -A INPUT -i lo -j ACCEPT
ip6tables -A OUTPUT -o lo -j ACCEPT
ip6tables -A INPUT -s ::1 -j DROP

# 3.4.2.3 Ensure IPv6 outbound and established connections are configured (Manual)
ip6tables -A OUTPUT -p tcp -m state --state NEW,ESTABLISHED -j ACCEPT
ip6tables -A OUTPUT -p udp -m state --state NEW,ESTABLISHED -j ACCEPT
ip6tables -A OUTPUT -p icmp -m state --state NEW,ESTABLISHED -j ACCEPT
ip6tables -A INPUT -p tcp -m state --state ESTABLISHED -j ACCEPT
ip6tables -A INPUT -p udp -m state --state ESTABLISHED -j ACCEPT
ip6tables -A INPUT -p icmp -m state --state ESTABLISHED -j ACCEPT



after implementing these my nginx ingress controller pod going in crashloopbackoff 
#Adding nodeport of nginx ingress controlleer 
iptables -I INPUT -p tcp -m tcp --dport 32443 -j ACCEPT  # For TLS traffic
iptables -I INPUT -p tcp -m tcp --dport 32002 -j ACCEPT  # For Health Checks
iptables -I INPUT -p tcp -m tcp --dport 32080 -j ACCEPT

is above rules which i added is correct or not?
plase help me into this

**Reply by arnaldo2792** [reply]
What's the error you are getting on your crashing pod? Do you have access to those logs?

**Reply by nike21oct** [reply]
my pod is getting failed with reason of liveness and readiness probe failed and it is because the iptables rule first drop all traffic and when we allow the node port on iptables it wont work and gets failed as my target group health is running on port 32002 and ingress nginx controller is listening on port 32443

**Reply by ozmodiar192** [reply]
It sounds like you're having a good ol' fashioned iptables issue.  The NodePort range is 30000-32767 so you can get any port in there.  This article may be helpful: https://kubernetes.io/blog/2023/05/11/nodeport-dynamic-and-static-allocation/#how-can-you-avoid-nodeport-service-port-conflicts

**Reply by nike21oct** [reply]
hi @ozmodiar192 , I did not understand your point how i can solve my issue . By reading the article it says about the range only but my issue is when I implement iptable rules to do the hardening of my bottlerocket AMI i am facing a issue.

**Reply by ozmodiar192** [reply]
I'm sorry, I may be misunderstanding your issue.  If you are creating services using nodeport and you've only allowed 3 ports within the node port range, it would make sense it's blocking if other ports get created.  Your options are to permit individual nodeports and then explicitly select the open nodeports, or allow a specific range of ports.  These may be helpful if that's the case, if not I apologize for misunderstanding the issue.
https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport
https://www.tkng.io/services/nodeport/

**Reply by nike21oct** [reply]
hi @ozmodiar192 , No this is not the issue you misunderstood  , the issue is I am having EKS cluster running by using AMI bottlerocket and I am doing the CIS benchmarking of bottlerocket AMI by following a document shared by AWS "https://aws.amazon.com/blogs/containers/validating-amazon-eks-optimized-bottlerocket-ami-against-the-cis-benchmark/" 
In these thee is a procedure of IP tables rules change and when I implement IP table rules my nginx ingress controller pod went to crashloopback off and I cannot access my application from outside the cluster.nginx as a ingress controller and when i implemented these IP tables rules by bootstrap-container my application stop open from outside the cluster , i mean my ingress is not functioning my nginx ingress controller pod went to crashloopbackoff , ngix controller load balancer in aws in target group the Protocol : Port is TLS: 32443 and health check is using protocol http and port is 32002, so what should i need to do here ?

**Comment by ozmodiar192** [comment]
I think you're assuming the CIS benchmarks are meant to handle your application traffic and they're not. They're meant to be a foundation on which to add the rules required for your applications. If you could put these rules in and it was easy and secure and no additional work was required, this would be the OOTB ruleset. It's not, because the default deny puts additional work on administrators to configure additional rules to permit their required traffic. Best of luck!
